### PR TITLE
draft: add web vitals collection

### PR DESCRIPTION
**Pull request by pauldambra**
if we gather web vitals info in clickhouse then we could do all kinds of wonderful magic

Screenshot 2024-05-31 at 17 13 25

pairs with https://github.com/PostHog/posthog-js/pull/1203

### CODE REVIEW

File: plugin-server/src/worker/ingestion/event-pipeline/extractWebVitalsDataStep.ts [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+import { castTimestampOrNow } from '../../../utils/utils'
+import { captureIngestionWarning, isNonEmptyString } from '../utils'
+import { EventPipelineRunner } from './runner'
+
+const validWebVitalNames = ['LCP', 'FCP', 'INP', 'CLS']
+
+interface ReceivedWebVitalEvent {
+    name: string
+    value: number
+    // we're not using the rest of the fields
```
Comment (pauldambra) [inline plugin-server/src/worker/ingestion/event-pipeline/extractWebVitalsDataStep.ts:12]: we're dumping info here e.g. events can be raised not on page navigation
we should probably also have a properties bag so we can filter in future or only store the events related to a  navigation for now

**Review by daibhin — APPROVED** [review]
Probably worth waiting for another set of 👀 on the CH changes but everything else looks good. One failing code check

File: plugin-server/src/config/kafka-topics.ts [hunk 2]
```diff
@@ -38,6 +38,8 @@ export const KAFKA_CLICKHOUSE_SESSION_REPLAY_EVENTS = `${prefix}clickhouse_sessi
 export const KAFKA_PERFORMANCE_EVENTS = `${prefix}clickhouse_performance_events${suffix}`
 // write heatmap events to ClickHouse
 export const KAFKA_CLICKHOUSE_HEATMAP_EVENTS = `${prefix}clickhouse_heatmap_events${suffix}`
+// write web vitals events to ClickHouse
+export const KAFKA_CLICKHOUSE_WEB_VITALS_EVENTS = `${prefix}clickhouse_web_vitals_events${suffix}`
```
Comment (pauldambra) [inline plugin-server/src/config/kafka-topics.ts:42]: TODO reminder to myself, these are autocreated in dev but would need to be terraformed in before merging this

File: posthog/hogql/database/schema/web_vitals.py [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+    fields: dict[str, FieldOrTable] = {
+        "session_id": StringDatabaseField(name="session_id"),
+        "team_id": IntegerDatabaseField(name="team_id"),
+        "timestamp": DateTimeDatabaseField(name="timestamp"),
+        "distinct_id": StringDatabaseField(name="distinct_id"),
+        "current_url": StringDatabaseField(name="current_url"),
+        "fcp": FloatDatabaseField(name="fcp"),
+        "lcp": FloatDatabaseField(name="lcp"),
+        "cls": FloatDatabaseField(name="cls"),
+        "inp": FloatDatabaseField(name="inp"),
```
Comment (pauldambra) [inline posthog/hogql/database/schema/web_vitals.py:21]: TODO need to add properties querying here

File: posthog/web_vitals/sql.py [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+    + """
+    PARTITION BY toYYYYMM(timestamp)
+    ORDER BY (team_id,  toDate(timestamp), current_url)
+    {ttl_period}
+"""
+).format(
+    table_name=WEB_VITALS_DATA_TABLE(),
+    cluster=settings.CLICKHOUSE_CLUSTER,
+    engine=WEB_VITALS_DATA_TABLE_ENGINE(),
+    ttl_period=ttl_period("timestamp", 366, unit="DAY"),
```
Comment (pauldambra) [inline posthog/web_vitals/sql.py:61]: this is IMHO a relatively long ttl but we could have a roll-up of older data if folk really want longer

File: posthog/web_vitals/sql.py
Comment (pauldambra) [inline posthog/web_vitals/sql.py:1]: @fuziontech ping, bump, bong 😊 

i only want to release this to team 2 in the short term (but equally don't want to do anything wilfully stupid 🙈)

Comment (pauldambra) [reply posthog/web_vitals/sql.py:1]: ping bump bong @fuziontech 

(volume here should be 1-5x pageview volume (for teams that have it enabled) if that helps)

Comment (fuziontech) [reply posthog/web_vitals/sql.py:1]: image

Comment (pauldambra) [reply posthog/web_vitals/sql.py:1]: does that count as approval @fuziontech 🤣 

(cc @robbie-c)

File: posthog/clickhouse/kafka_engine.py [hunk 5]
```diff
@@ -42,7 +42,12 @@ def kafka_engine(topic: str, kafka_host: str | None = None, group="group1") -> s
     return KAFKA_ENGINE.format(topic=topic, kafka_host=kafka_host, group=group, serialization="JSONEachRow")
 
 
-def ttl_period(field: str = "created_at", amount: int = 3, unit: Literal["DAY", "WEEK"] = "WEEK") -> str:
+def ttl_period(
+    field: str = "created_at",
+    amount: int = 3,
+    # support a subset of valid options for clickhouse interval
+    unit: Literal["DAY", "WEEK", "MONTH", "YEAR"] = "WEEK",
+) -> str:
```
Comment (fuziontech) [inline posthog/clickhouse/kafka_engine.py:50]: 👍

File: posthog/web_vitals/sql.py [hunk 6]
```diff
@@ -0,0 +1,125 @@
...
+"""
+
+WEB_VITALS_DATA_TABLE_ENGINE = lambda: MergeTreeEngine("web_vitals", replication_scheme=ReplicationScheme.SHARDED)
+
+WEB_VITALS_TABLE_SQL = lambda: (
+    WEB_VITALS_TABLE_BASE_SQL
+    + """
+    PARTITION BY toYYYYMM(timestamp)
+    ORDER BY (team_id,  toDate(timestamp), current_url)
+    {ttl_period}
```
Comment (fuziontech) [inline posthog/web_vitals/sql.py:55]: 👏 nice order by

**Review by fuziontech — APPROVED** [review]
ClickHouse bits look good! Just keep in mind we can't keep adding Kafka tables forever...sadly.

**Comment by pauldambra** [comment]
related to https://github.com/PostHog/posthog/issues/23045

**Pull request closed without merging**